packages/commonwealth/server/util/roles.ts

```diff
@@ -0,0 +1,262 @@
+      deny: this.deny,
+    };
+  }
+}
+
+export async function findAllRoles(
+  models: DB,
+  findOptions: FindOptions,
+  chain_id?: string,
+  permissions?: Permission[]
+): Promise<RoleInstanceWithPermission[]> {
+  let roleFindOptions: any;
+  const whereCondition = {};
+  if (chain_id) {
+    whereCondition['chain_id'] = chain_id;
+  }
+  if (permissions) {
+    roleFindOptions = {
+      where: {
+        [Op.and]: [
+          { chain_id },
+          {
+            name: {
+              [Op.or]: permissions,
+            },
+          },
+        ],
+      },
+      include: [
+        {
+          model: models.RoleAssignment,
+          ...findOptions,
+        },
+      ],
+    };
+  } else {
+    roleFindOptions = {
+      where: whereCondition,
+      include: [
+        {
+          model: models.RoleAssignment,
+          ...findOptions,
+        },
+      ],
+    };
+  }
+  const communityRoles = await models.CommunityRole.findAll(roleFindOptions);
+  const roles: RoleInstanceWithPermission[] = [];
+  if (communityRoles) {
+    for (const communityRole of communityRoles) {
+      const roleAssignments = communityRole.RoleAssignments;
+      if (roleAssignments.length > 0) {
+        for (const roleAssignment of roleAssignments) {
+          const role = new RoleInstanceWithPermission(
+            roleAssignment.toJSON(),
+            communityRole.chain_id,
+            communityRole.name,
+            communityRole.allow,
+            communityRole.deny
+          );
+          roles.push(role);
+        }
+      }
+    }
+  } else {
+    return null;
+  }
+  return roles;
+}
+
+export async function findOneRole(
+  models: DB,
+  findOptions: FindOptions,
+  chain_id: string,
+  permissions?: Permission[]
+): Promise<RoleInstanceWithPermission> {
+  let roleFindOptions: any;
+  if (permissions) {
+    roleFindOptions = {
+      where: {
+        [Op.and]: [
+          { chain_id },
+          {
+            name: {
+              [Op.or]: permissions,
+            },
+          },
+        ],
+      },
+      include: [
+        {
+          model: models.RoleAssignment,
+          ...findOptions,
+        },
+      ],
+      order: [
+        sequelize.literal(
+          `CASE WHEN name='admin' THEN 0 WHEN name='moderator' THEN 1 ELSE 2 END`
+        ),
+      ],
+    };
+  } else {
+    roleFindOptions = {
+      where: {
+        chain_id,
+      },
+      include: [
+        {
+          model: models.RoleAssignment,
+          ...findOptions,
+        },
+      ],
+      order: [
+        sequelize.literal(
+          `CASE WHEN name='admin' THEN 0 WHEN name='moderator' THEN 1 ELSE 2 END`
+        ),
+      ],
+    };
+  }
+
+  const communityRole = await models.CommunityRole.findOne(roleFindOptions);
+  let role: RoleInstanceWithPermission;
+  if (communityRole) {
+    // Retrieve the highest role as it will be the highest permission role for the address_id
+    if (communityRole.RoleAssignments.length > 0) {
```

Comment:
We should factor the conversion from multiple RoleAssignments -> one RoleInstanceWithPermission into a helper function, so we can unit test it. We need to make sure that conversion works + that RoleInstanceWithPermission looks identical to an old Role type, to ensure against regressions.